### PR TITLE
Use oEmbed for YouTube Shorts

### PR DESCRIPTION
**Pull request by SuperSandro2000**
Based on https://core.trac.wordpress.org/changeset/53106

### Pull Request Checklist

* [x] Pull request is based on the develop branch
* [x] Pull request includes a [changelog file](https://matrix-org.github.io/synapse/latest/development/contributing_guide.html#changelog). The entry should:
  - Be a short description of your change which makes sense to users. "Fixed a bug that prevented receiving messages from other servers." instead of "Moved X method from `EventStore` to `EventWorkerStore`.".
  - Use markdown where necessary, mostly for `code blocks`.
  - End with either a period (.) or an exclamation mark (!).
  - Start with a capital letter.
  - Feel free to credit yourself, by adding a sentence "Contributed by @github_username." or "Contributed by [Your Name]." to the end of the entry.
* [x] Pull request includes a [sign off](https://matrix-org.github.io/synapse/latest/development/contributing_guide.html#sign-off)
* [x] [Code style](https://matrix-org.github.io/synapse/latest/code_style.html) is correct
  (run the [linters](https://matrix-org.github.io/synapse/latest/development/contributing_guide.html#run-the-linters))

### CODE REVIEW

**Comment by clokep** [comment]
Do you have a before / after of the result of what gets changed by including this? (And an example URL?)

**Comment by SuperSandro2000** [comment]
Before it displayed the youtube login page and now the preview like for any other youtube. For testing any shorts url suffices. 

image

**Comment by MatMaul** [comment]
Would it be possible to test with the official Youtube oEmbed descriptor from [here](https://oembed.com/providers.json) ?

If it works well I think we should rather use this one.

Is it possible we crossed each other at FOSDEM recently ? I talked with someone about that, that may have been you, or not :)

**Review by MatMaul — COMMENTED** [review]
Would it be possible to test with the official Youtube oEmbed descriptor from [here](https://oembed.com/providers.json) ?

If it works well I think we should rather use this one.

Is it possible we crossed each other at FOSDEM recently ? I talked with someone about that, that may have been you, or not :)

**Comment by clokep** [comment]
> Would it be possible to test with the official Youtube oEmbed descriptor from [here](https://oembed.com/providers.json) ?

We would want to double check it doesn't regress anything on those other endpoints.

**Comment by SuperSandro2000** [comment]
> Would it be possible to test with the official Youtube oEmbed descriptor from [here](https://oembed.com/providers.json) ?

The only differences is that mine also adds http and theirs add discovery: true. I don't know oembed if that makes a difference but it sounds like it should not to me.

> Is it possible we crossed each other at FOSDEM recently ? I talked with someone about that, that may have been you, or not :)

Not, it's not.

**Comment by clokep** [comment]
> I tried adding the oembed providers.json and ran into the following problems:

I think we don't want to include the whole file, there are some providers which give much worse previews with oEmbed (we do use autodiscovery though 🤷 ).

**Comment by SuperSandro2000** [comment]
So, whats happening with this PR now?

**Comment by clokep** [comment]
> So, whats happening with this PR now?

I think it just never got added back to the queue for some reason...

**Comment by clokep** [comment]
> Before it displayed the youtube login page and now the preview like for any other youtube. For testing any shorts url suffices.

For testing I used https://www.youtube.com/shorts/usofnFoFNN4

Before this branch I got:

image

<details>
<summary>Open Graph response</summary>

```json
{
  "og:site_name": "YouTube",
  "og:url": "https://www.youtube.com/shorts/usofnFoFNN4",
  "og:title": "Meet Kevin Bacon, the dog who grabs a beer for you! #JimmyFallon #FallonTonight",
  "og:image:width": 480,
  "og:image:height": 360,
  "og:type": "video.other",
  "og:video:url": "https://www.youtube.com/embed/usofnFoFNN4",
  "og:video:secure_url": "https://www.youtube.com/embed/usofnFoFNN4",
  "og:video:type": "text/html",
  "og:video:width": 113,
  "og:video:height": 200,
  "og:app:name:iphone": "YouTube",
  "og:app:id:iphone": "544007664",
  "og:app:name:ipad": "YouTube",
  "og:app:id:ipad": "544007664",
  "og:app:url:iphone": "vnd.youtube://www.youtube.com/shorts/usofnFoFNN4?feature=applinks",
  "og:app:url:ipad": "vnd.youtube://www.youtube.com/shorts/usofnFoFNN4?feature=applinks",
  "og:app:name:googleplay": "YouTube",
  "og:app:id:googleplay": "com.google.android.youtube",
  "og:app:url:googleplay": "https://www.youtube.com/shorts/usofnFoFNN4",
  "og:player": "https://www.youtube.com/embed/usofnFoFNN4",
  "og:player:width": "405",
  "og:player:height": "720",
  "og:description": "About\n\nPress\n\nCopyright\n\nContact us\n\nCreators\n\nAdvertise\n\nDevelopers\n\nTerms\n\nPrivacy\n\nPolicy & Safety\n\nHow YouTube works\n\nTest new features\n\nNFL Sunday Ticket\n\n© 2023 Google LLC",
  "og:image": "http://localhost:8080/_matrix/media/v3/thumbnail/localhost:8480/2023-04-10_LlaqneEPkPXgudwk?height=96&width=96",
  "og:image:type": "image/jpeg",
  "matrix:image:size": 8993
}
```

</details>

On this branch I got:

image

<details>
<summary>Open Graph response</summary>

```json
{
  "og:url": "https://www.youtube.com/shorts/usofnFoFNN4",
  "og:title": "Meet Kevin Bacon, the dog who grabs a beer for you! #JimmyFallon #FallonTonight",
  "og:site_name": "YouTube",
  "og:type": "video.other",
  "og:video:width": 113,
  "og:video:height": 200,
  "og:image:width": 480,
  "og:image:height": 360,
  "og:image": "http://localhost:8080/_matrix/media/v3/thumbnail/localhost:8480/2023-04-10_pVipqwdHrIpbvpnE?height=96&width=96",
  "og:image:type": "image/jpeg",
  "matrix:image:size": 8993
}
```

----

The differences I see are:

* Without this change we get a garbage `description` field. ✅ 
* With this change the Open Graph response loses a bunch of info (in particular the *true* embed URL). ❌ 

But it sounds like you don't see *any* response without the oEmbed configuration? I wonder if this is a difference between region (I'm in the US).

I'm unsure if any clients actually use the video bits of the Open Graph response, unfortunately. 😢

**Comment by SuperSandro2000** [comment]
My homeserver is located in Germany and I only get the below picture for the link you posted.

image

I am not sure how to generate the oembed result without youtube's oembed url.

```
 ➜ curl -s "https://www.youtube.com/oembed/?format=json&url=https://www.youtube.com/shorts/usofnFoFNN4" | jq .
{
  "author_name": "The Tonight Show Starring Jimmy Fallon",
  "author_url": "https://www.youtube.com/@fallontonight",
  "height": 200,
  "html": "<iframe width=\"200\" height=\"113\" src=\"https://www.youtube.com/embed/usofnFoFNN4?feature=oembed\" frameborder=\"0\" allow=\"accelerometer; autoplay; clipboard-write; encrypted-media; gyroscope; picture-in-picture; web-share\" allowfullscreen title=\"Meet Kevin Bacon, the dog who grabs a beer for you! #JimmyFallon #FallonTonight\"></iframe>",
  "provider_name": "YouTube",
  "provider_url": "https://www.youtube.com/",
  "thumbnail_height": 360,
  "thumbnail_url": "https://i.ytimg.com/vi/usofnFoFNN4/hq2.jpg",
  "thumbnail_width": 480,
  "title": "Meet Kevin Bacon, the dog who grabs a beer for you! #JimmyFallon #FallonTonight",
  "type": "video",
  "version": "1.0",
  "width": 113
}
```

**Review by clokep — APPROVED** [review]

**Comment by clokep** [comment]
Sorry for the delay on this, sounds like there's general agreement with the above so I'll mash the button.

File: changelog.d/15025.misc [hunk 1]
```diff
@@ -0,0 +1 @@
+Add oembed providers for youtube shorts.
```
Comment (clokep) [inline changelog.d/15025.misc:1]:
```suggestion
Use oEmbed to generate URL previews for YouTube Shorts.
```

**Comment by SuperSandro2000** [comment]
Thanks!

**Pull request merged**